lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py

```diff
@@ -1,31 +1,106 @@
+        assert exitcode == 0
+
+
+class Test_list_tools_tools_registered:
+    @pytest.fixture
+    def tool(self, pbench_run):
+        p = pbench_run / "tools-v1-default" / "testhost.example.com"
+        p.mkdir(parents=True)
+        tool = p / "perf"
+        tool.touch()
+
+    @pytest.fixture
+    def tool_with_random_file(self, pbench_run):
+        p = pbench_run / "tools-v1-default" / "testhost.example.com"
+        p.mkdir(parents=True)
+        (p.parent / "foo").touch()
+        tool = p / "perf"
+        tool.touch()
+
+    def test_help(self, tool, agent_config):
+        command = ["pbench-list-tools", "--help"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert traceback not in err
+        assert empty == err
+        assert usage in out
+        assert exitcode == 0
+
+    def test_no_args(self, tool, agent_config):
+        command = ["pbench-list-tools"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert traceback not in err
+        assert empty == err
+        assert b"default: testhost.example.com ['perf']" in out
+        assert exitcode == 0
+
+    def test_name(self, tool, agent_config):
+        command = ["pbench-list-tools", "-n", "perf"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert traceback not in err
+        assert empty == err
+        assert b"tool name: perf groups: default" in out
+        assert exitcode == 0
+
+    # Issue #2345
+    def test_name_with_random_file(self, tool_with_random_file, agent_config):
+        command = ["pbench-list-tools", "-n", "perf"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert traceback not in err
+        assert empty == err and exitcode == 0
+        assert b"tool name: perf groups: default" in out
+
+    # Issue #2302
+    def test_unknown_group(self, tool, agent_config):
+        command = ["pbench-list-tools", "-g", "unknown"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert traceback not in err
+        assert bad_group in err
+        assert empty == out
+        # this is 0.69.9 behavior
+        assert exitcode == 0
+
+
+class Test_list_tools_tools_registered_with_options:
+    @pytest.fixture
+    def single_group_tools(self, pbench_run):
+        p = pbench_run / "tools-v1-default" / "testhost.example.com"
+        p.mkdir(parents=True)
+        tool = p / "iostat"
+        tool.write_text("--interval=30")
+        tool = p / "mpstat"
+        tool.write_text("--interval=300")
+
+    # Issue #2346
+    def test_existing_group_options(self, single_group_tools, agent_config):
+        command = ["pbench-list-tools", "--with-option"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert traceback not in err
+        assert empty == err and exitcode == 0
+        assert (
+            b"default: testhost.example.com [('iostat', '--interval=30'), ('mpstat', '--interval=300')]\n"
```

Comment:
I'm looking forward to having more human readable output....

Reply:
Coming in the subsequent PR - stay tuned.